### PR TITLE
Export XDG_DATA_DIRS for custom installation paths

### PR DESCRIPTION
**Pull request by hadess**
See #2032

### CODE REVIEW

File: app/flatpak-main.c [hunk 1]
```diff
@@ -146,6 +147,7 @@ static GOptionEntry empty_entries[] = {
   { "default-arch", 0, 0, G_OPTION_ARG_NONE, &opt_default_arch, N_("Print default arch and exit"), NULL },
   { "supported-arches", 0, 0, G_OPTION_ARG_NONE, &opt_supported_arches, N_("Print supported arches and exit"), NULL },
   { "gl-drivers", 0, 0, G_OPTION_ARG_NONE, &opt_gl_drivers, N_("Print active gl drivers and exit"), NULL },
+  { "list-installations", 0, 0, G_OPTION_ARG_NONE, &opt_list_installations, N_("Print the paths for system installations and exit"), NULL },
```
Comment (matthiasclasen) [inline app/flatpak-main.c]: This docstring should perhaps say "custom installations" ?

To make it clear that this does not include the default system location

Comment (hadess) [reply app/flatpak-main.c]: It shows all the installation directories, including the default system one.

**Comment by aiamuzz** [comment]
does it mean this will fix the issue i have been facing with regard to the missing icons(.desktop fles) in the launcher ... for apps installed at a custom location ?

File: app/flatpak-main.c (hunk 1 above)
Comment (matthiasclasen) [reply app/flatpak-main.c]: oh, I must have misread, then

**Comment by matthiasclasen** [comment]
Yes, it will

**Comment by hadess** [comment]
After fixing the shell script to not pipe out (which creates a sub-shell, and my `XDG_DATA_DIRS` values didn't have a way to get back out to the original shell) those are the test results:

XDG_DATA_DIRS value in jhbuild (installs under `~/Projects/gnome-install`):
```
$ cat ~/Projects/gnome-install/etc/flatpak/installations.d/sd-card.conf
[Installation "sd-card"]
Path=/opt/flatpak-test
DisplayName=SD card
$ ./flatpak --list-installations 
/opt/flatpak-test
/home/hadess/Projects/gnome-install/var/lib/flatpak
$ export XDG_DATA_DIRS=/home/hadess/Projects/gnome-install/share:/usr/local/share/:/usr/share/
$ . flatpak.sh 
$ echo $XDG_DATA_DIRS 
/home/hadess/.local/share/flatpak/exports/share:/opt/flatpak-test/exports/share:/home/hadess/Projects/gnome-install/var/lib/flatpak/exports/share:/home/hadess/Projects/gnome-install/share:/usr/local/share/:/usr/share/
```

And using the system flatpak (which doesn't have the custom installation path):
```
$ export XDG_DATA_DIRS=/home/hadess/Projects/gnome-install/share:/usr/local/share/:/usr/share/
$ . /etc/profile.d/flatpak.sh 
$ echo $XDG_DATA_DIRS
/home/hadess/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/home/hadess/Projects/gnome-install/share:/usr/local/share/:/usr/share/
```

The diff between the 2:
```patch
-/var/lib/flatpak/exports/share
+/opt/flatpak-test/exports/share
+/home/hadess/Projects/gnome-install/var/lib/flatpak/exports/share
```

Default system path is replaced (jhbuild vs system), and put after the custom installation paths, so seems to work!

**Comment by alexlarsson** [comment]
This will only work for the profile snippet though, not the other ways we're setting the env vars. And the profile is really not good enough. For instance, it generally runs after the dbus daemon is launched, so dbus will not read service files from there.

**Comment by matthiasclasen** [comment]
Where are those other ways ?

**Comment by hadess** [comment]
> This will only work for the profile snippet though, not the other ways we're setting the env vars. And the profile is really not good enough. For instance, it generally runs after the dbus daemon is launched, so dbus will not read service files from there.

Isn't that what `dbus-update-activation-environment` should be used for?

**Comment by hadess** [comment]
> The activation environment is the environment for the activated services, I believe.

Making D-Bus daemon update what it thinks are directories it needs to watch doesn't seem too complicated. In `bus_driver_handle_update_activation_environment()` (in `bus/driver.c`), it would catch the special case `XDG_DATA_DIRS`, update its `watched_dirs`, and reload the `.service` files there.

@smcv What do you think?

**Comment by alexlarsson** [comment]
We also modify it in `env.d/flatpak.env.in` which is set by gdm, and in `dbus.service.d/flatpak.conf.in` + `system-helper/flatpak-system-helper.service.in` which are set by systemd. Some packages also have other methods.

Basically this whole area is a gigantic pain in the ass. @smcv talked about a better way of doing this, using some kind of pam-like setup. I don't know if anything ever came out of that.

**Comment by matthiasclasen** [comment]
I found:
/usr/lib/systemd/user/dbus.service.d/flatpak.conf
/usr/share/gdm/env.d/flatpak.env

**Comment by matthiasclasen** [comment]
Are you saying that env.d should be sufficient for the session bus, and we don't need the systemd shenanigans anymore ?

**Comment by matthiasclasen** [comment]
closing this in favor of my other pr

**Comment by hadess** [comment]
> closing this in favor of my other pr

Would be nice to add a link to the other PR in question:
https://github.com/flatpak/flatpak/pull/2122

**Comment by matthiasclasen** [comment]
sorry, didn't have the link at hand

**Pull request closed without merging**